man/RNAplfold.1
2.1.7

.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.38.2.
.TH RNAPLFOLD "1" "April 2014" "RNAplfold 2.1.7" "User Commands"
.SH NAME
RNAplfold \- manual page for RNAplfold 2.1.7
.SH SYNOPSIS
.B RNAplfold
[\fIOPTIONS\fR]...
.SH DESCRIPTION
RNAplfold 2.1.7
.PP
calculate locally stable secondary structure \- pair probabilities
.PP
Computes local pair probabilities for base pairs with a maximal span of L. The
probabilities are averaged over all windows of size L that contain the base
pair. For a sequence of length n and a window size of L the algorithm uses only
O(n+L*L) memory and O(n*L*L) CPU time. Thus it is practical to "scan" very
large genomes for short stable RNA structures.
.PP

Output consists of a dot plot in postscript file, where the averaged pair probabilities
can easily be parsed and visually inspected.

The -u option makes i possible to compute the probability that a stretch of x consequtive
nucleotides is unpaired, which is useful for predicting possible binding sites. Again
this probability is averaged over all windows containing the region.

.B WARNING! Output format changed!!

The output is a plain text matrix containing on each line a position i followed by the
probability that i is unpaired, [i-1..i] is unpaired [i-2..i] is unpaired and so on to
the probability that [i-x+1..i] is unpaired.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-\-detailed\-help\fR
Print help, including all details and hidden
options, and exit
.TP
\fB\-\-full\-help\fR
Print help, including hidden options, and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.SS "General Options:"
.IP
Below are command line options which alter the general behavior of this
program
.TP
\fB\-W\fR, \fB\-\-winsize\fR=\fIsize\fR
Average the pair probabilities over windows of
given size
.IP
(default=`70')
.TP
\fB\-L\fR, \fB\-\-span\fR=\fIsize\fR
Set the maximum allowed separation of a base pair
to span. I.e. no pairs (i,j) with j\-i > span
will be allowed. Defaults to winsize if
parameter is omitted
.TP
\fB\-c\fR, \fB\-\-cutoff\fR=\fIFLOAT\fR
Report only base pairs with an average probability
> cutoff in the dot plot
.IP
(default=`0.01')
.TP
\fB\-o\fR, \fB\-\-print_onthefly\fR
Save memory by printing out everything during
computation.
NOTE: activated per default for sequences over
1M bp.
.IP
(default=off)
.TP
\fB\-u\fR, \fB\-\-ulength\fR=\fIlength\fR
Compute the mean probability that regions of
length 1 to a given length are unpaired. Output
is saved in a _lunp file.
.IP
(default=`31')
.TP
\fB\-O\fR, \fB\-\-opening_energies\fR
Switch output from probabilities to their
logarithms, which are NOT exactly the mean
energies needed to the respective stretch of
bases!
NOTE: This actives \fB\-u\fR option.
.IP
(default=off)
.TP
\fB\-\-plex_output\fR
Create additional output files for RNAplex.
.IP
(default=off)
.TP
\fB\-\-noconv\fR
Do not automatically substitude nucleotide "T"
with "U"
.IP
(default=off)
.SS "Model Details:"
.TP
\fB\-T\fR, \fB\-\-temp\fR=\fIDOUBLE\fR
Rescale energy parameters to a temperature of temp
C. Default is 37C.
.TP
\fB\-4\fR, \fB\-\-noTetra\fR
Do not include special tabulated stabilizing
energies for tri\-, tetra\- and hexaloop hairpins.
Mostly for testing.
.IP
(default=off)
.TP
\fB\-d\fR, \fB\-\-dangles\fR=\fIINT\fR
How to treat "dangling end" energies for bases
adjacent to helices in free ends and multi\-loops
.IP
(default=`2')
.IP
With \fB\-d2\fR dangling energies will be added for the bases adjacent to a helix on
both sides in any case.
.HP
\fB\-d0\fR ignores dangling ends altogether (mostly for debugging).
.TP
\fB\-\-noLP\fR
Produce structures without lonely pairs (helices
of length 1).
.IP
(default=off)
.IP
For partition function folding this only disallows pairs that can only occur
isolated. Other pairs may still occasionally occur as helices of length 1.
.TP
\fB\-\-noGU\fR
Do not allow GU pairs
.IP
(default=off)
.TP
\fB\-\-noClosingGU\fR
Do not allow GU pairs at the end of helices
.IP
(default=off)
.TP
\fB\-P\fR, \fB\-\-paramFile\fR=\fIparamfile\fR
Read energy parameters from paramfile, instead of
using the default parameter set.
.IP
A sample parameter file should accompany your distribution.
See the RNAlib documentation for details on the file format.
.TP
\fB\-b\fR, \fB\-\-binaries\fR
Output accessibility profiles in binary format
\&.  (default=off)
.IP
The binary files produced by RNAplfold do not need to be parsed by RNAplex,
.IP
so that they are directly loaded into memory. This is useful when large
sequences have to be searched for putative hybridization sites. An other
advantage of the binary format is the 50% file size decrease.
.TP
\fB\-\-nsp\fR=\fISTRING\fR
Allow other pairs in addition to the usual
AU,GC,and GU pairs.
.IP
Its argument is a comma separated list of additionally allowed pairs. If the
first character is a "\-" then AB will imply that AB and BA are allowed
pairs.
e.g. RNAfold \fB\-nsp\fR \fB\-GA\fR  will allow GA and AG pairs. Nonstandard pairs are
given 0 stacking energy.
.TP
\fB\-e\fR, \fB\-\-energyModel\fR=\fIINT\fR
Rarely used option to fold sequences from the
artificial ABCD... alphabet, where A pairs B,
C\-D etc.  Use the energy parameters for GC (\fB\-e\fR
1) or AU (\fB\-e\fR 2) pairs.
.TP
\fB\-\-betaScale\fR=\fIDOUBLE\fR
Set the scaling of the Boltzmann factors
(default=`1.')
.IP
The argument provided with this option enables to scale the thermodynamic
temperature used in the Boltzmann factors independently from the temperature
used to scale the individual energy contributions of the loop types. The
Boltzmann factors then become exp(\fB\-dG\fR/(kT*betaScale)) where k is the
Boltzmann constant, dG the free energy contribution of the state and T the
absolute temperature.
.SH AUTHOR

Stephan H Bernhart, Ivo L Hofacker, Peter F Stadler, Ronny Lorenz
.SH REFERENCES
.I If you use this program in your work you might want to cite:

R. Lorenz, S.H. Bernhart, C. Hoener zu Siederdissen, H. Tafer, C. Flamm, P.F. Stadler and I.L. Hofacker (2011),
"ViennaRNA Package 2.0",
Algorithms for Molecular Biology: 6:26 

I.L. Hofacker, W. Fontana, P.F. Stadler, S. Bonhoeffer, M. Tacker, P. Schuster (1994),
"Fast Folding and Comparison of RNA Secondary Structures",
Monatshefte f. Chemie: 125, pp 167-188


S. H. Bernhart, I.L. Hofacker, and P.F. Stadler (2006),
"Local Base Pairing Probabilities in Large RNAs",
Bioinformatics: 22, pp 614-615

A.F. Bompfuenewerer, R. Backofen, S.H. Bernhart, J. Hertel, I.L. Hofacker, P.F. Stadler, S. Will (2007),
"Variations on RNA Folding and Alignment: Lessons from Benasque",
J. Math. Biol.


.I The energy parameters are taken from:

D.H. Mathews, M.D. Disney, D. Matthew, J.L. Childs, S.J. Schroeder, J. Susan, M. Zuker, D.H. Turner (2004),
"Incorporating chemical modification constraints into a dynamic programming algorithm for prediction of RNA secondary structure",
Proc. Natl. Acad. Sci. USA: 101, pp 7287-7292

D.H Turner, D.H. Mathews (2009),
"NNDB: The nearest neighbor parameter database for predicting stability of nucleic acid secondary structure",
Nucleic Acids Research: 38, pp 280-282
.SH "REPORTING BUGS"
If in doubt our program is right, nature is at fault.
.br
Comments should be sent to rna@tbi.univie.ac.at.
.SH "SEE ALSO"

RNALfold(1)


NEWS
Some Changes relative to ViennaRNA-1.03

New energy parameters used in the folding routines.
Prediction of minimum free energy structures now includes contributions
from terminal unpaired nucleotides (so called dangling ends).
The partition function treats these contributions approximately.

Energies for GU pairs in are now taken from
L. He et al., Biochemistry 1991, 30, p 11124

Bulges of length 1 now get an additional stacking energy.

The folding routines accept  a list of nonstandard basepairs that should be
allowed to occur. Also, when doing constrained folding, forced pairs will
be honoured even if the pair is usally not allowed (not AU, GC, or GU).
RNAeval should be useful now for structures containing nonstandard pairs.

Some support for constrained folding using partition function: A constraint
of the form i-j must pair or i must not pair will be honoured by disallowing 
all pairs conflicting with the constraint.

Both folding algorthms use less memory and should be a bit faster for long
sequences.

Several smaller bug fixes.


Progs/RNAplfold.c
/* Last changed Time-stamp: <2006-08-09 16:05:15 berni> */
/*
		  Ineractive Access to folding Routines

		  c Ivo L Hofacker
		  Vienna RNA package
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <unistd.h>
#include <string.h>
#include "fold.h"
#include "part_func.h"
#include "fold_vars.h"
#include "utils.h"
#include "PS_dot.h"
#include "LPfold.h"

extern float Lfold(char *string, char *structure, int winsize);
extern void  read_parameter_file(const char fname[]);

/*@unused@*/
static char rcsid[] = "$Id: RNAplfold.c,v 1.5 2007/02/02 15:14:27 ivo Exp $";

#define PRIVATE static

static char  scale[] = "....,....1....,....2....,....3....,....4"
	"....,....5....,....6....,....7....,....8";

PRIVATE void usage(void);
PRIVATE void putout_pup(float *pup,int length, int winsize);
int unpaired;
/*--------------------------------------------------------------------------*/

int main(int argc, char *argv[])
{
  char *string, *line;
  char *structure=NULL, *cstruc=NULL;
  char  fname[30], ffname[20];
  char  *ParamFile=NULL;
  char  *ns_bases=NULL, *c;
  int   i, length, l, sym,r;
  int   istty;
  int noconv=0;
  int winsize=70;
  int pairdist=0;
  float cutoff=0.01;
  int hit;
  float *pup=NULL; /*prob of being unpaired*/
  plist *pl;
  int
  do_backtrack = 1;
  string=NULL;
  dangles=2;
   unpaired=0;
   for (i=1; i<argc; i++) {
    if (argv[i][0]=='-')
      switch ( argv[i][1] )
	{
	case 'T':  if (argv[i][2]!='\0') usage();
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i], "%lf", &temperature);
	  if (!r) usage();
	  break;
	case 'n':
	  if ( strcmp(argv[i], "-noGU")==0) noGU=1;
	  if ( strcmp(argv[i], "-noCloseGU")==0) no_closingGU=1;
	  if ( strcmp(argv[i], "-noLP")==0) noLonelyPairs=1;
	  if ( strcmp(argv[i], "-nsp") ==0) {
	    if (i==argc-1) usage();
	    ns_bases = argv[++i];
	  }
	  if ( strcmp(argv[i], "-noconv")==0) noconv=1;
	  break;
	case '4':
	  tetra_loop=0;
	  break;
	case 'e':
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i],"%d", &energy_set);
	  if (!r) usage();
	  break;
	case 'c':
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i],"%f", &cutoff);
	  if (!r) usage();
	  break;
/*	case 'C':
	  fold_constrained=1;
	  break;
	case 'S':
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i],"%lf", &sfact);
	  if (!r) usage();
	  break;
*/
	case 'd': dangles=0;
	  if (argv[i][2]!='\0') {
	    r=sscanf(argv[i]+2, "%d", &dangles);
	    if (r!=1) usage();
	    if (!dangles%2) usage(); /*1,3 not possible*/
	  }
	  break;
	case 'P':
	  if (i==argc-1) usage();
	  ParamFile = argv[++i];
	  break;
	case 'W':
	  if (i==argc-1) usage();
	  r=sscanf(argv[++i], "%d", &winsize);
	  if (r!=1) usage();
	  break;
	case 'u':
	  if (i==argc-1) usage();
	  r=sscanf(argv[++i], "%d", &unpaired);
	  if (r!=1) usage();
	  if (unpaired<0) usage();
	  break;
	case 'L':
	  if (i==argc-1) usage();
	  r=sscanf(argv[++i], "%d", &pairdist);
	  if (r!=1) usage();
	  break;
	default: usage();
	}
  }

  if (ParamFile != NULL)
    read_parameter_file(ParamFile);

  if (ns_bases != NULL) {
    nonstandards = space(33);
    c=ns_bases;
    i=sym=0;
    if (*c=='-') {
      sym=1; c++;
    }
    while (*c!='\0') {
      if (*c!=',') {
	nonstandards[i++]=*c++;
	nonstandards[i++]=*c;
	if ((sym)&&(*c!=*(c-1))) {
	  nonstandards[i++]=*c;
	  nonstandards[i++]=*(c-1);
	}
      }
      c++;
    }
  }
  istty = isatty(fileno(stdout))&&isatty(fileno(stdin));
   if (pairdist==0) pairdist=winsize;
  if (pairdist>winsize) {
    fprintf(stderr, "pairdist (-L %d) should be <= winsize (-W %d);"
	    "Setting pairdist=winsize\n",pairdist, winsize);
    pairdist=winsize;
  }

  do {				/* main loop: continue until end of file */
    if (istty) {
      printf("\nInput string (upper or lower case); @ to quit\n");
      printf("%s\n", scale);
    }
    fname[0]='\0';
    if ((line = get_line(stdin))==NULL) break;

    /* skip comment lines and get filenames */
    while ((*line=='*')||(*line=='\0')||(*line=='>')) {
      if (*line=='>')
	(void) sscanf(line, ">%12s", fname);
      printf("%s\n", line);
      free(line);
      if ((line = get_line(stdin))==NULL) break;
    }

    if ((line ==NULL) || (strcmp(line, "@") == 0)) break;

    string = (char *) space(strlen(line)+1);
    (void) sscanf(line,"%s",string);
    free(line);
    length = (int) strlen(string);
    if (unpaired) {
      pup=(float *)space((length+1)*sizeof(float));
      pup[0]=unpaired;
    }

    structure = (char *) space((unsigned) length+1);

    for (l = 0; l < length; l++) {
      string[l] = toupper(string[l]);
      if (!noconv && string[l] == 'T') string[l] = 'U';
    }
    if (istty)
      printf("length = %d\n", length);

    /* initialize_fold(length); */
    update_fold_params(); 
    if (length<winsize) {
      fprintf(stderr, "WARN: window size %d larger than sequence length %d\n",
	      winsize, length);
      winsize=length;
    }
    if (length >= 5) {
      pf_scale = -1;
   
      pl=pfl_fold(string, winsize, pairdist, cutoff, pup);
      if (fname[0]!='\0') {
	strcpy(ffname, fname);
	strcat(ffname, "_dp.ps");
      }
      else strcpy(ffname, "plfold_dp.ps");
      PS_dot_plot_turn(string, pl, ffname, pairdist);
      free(pl);
      if (unpaired) {
	putout_pup(pup,length,winsize);
	free(pup);
      }
      if (cstruc!=NULL) free(cstruc);
    (void) fflush(stdout);
    }
    free(string);
    free(structure);
  } while (1);
  return 0;
}

PRIVATE void usage(void)
{
  nrerror("usage:\n"
	  "RNAplfold [-L span] [-W winsize] [-u size of unpaired region]\n"
	  "          [-T temp] [-4] [-d[0|1|2]] [-noGU] [-noCloseGU]\n"
	  "          [-noLP] [-P paramfile] [-nsp pairs] [-noconv]\n");
}

PRIVATE void putout_pup(float *pup,int length, int winsize) {
  int i;
  float factor;
  float tfact;
  printf("# prob of being unpaired between i-%d and i\n",unpaired);
  fflush(NULL);
  for (i=unpaired; i<=length; i++) {
    if (i<winsize) {
      factor=1./(i-unpaired+1);
    }
    if (i>length-winsize+unpaired-1) {
      tfact=1./(length-i+1);
      if (tfact>factor) {
	factor=tfact;
      }
     
    }
    else {
      tfact=1./(winsize-unpaired+1);
      if (tfact>factor) {
	factor=tfact;
      }
    }
    printf("%d %.6f\n",i,pup[i]*factor);
  }


}
